File: cpp/lc-695.cpp
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int ans = 0;
        vector<vector<bool>> visited(grid.size(), vector<bool>(grid[0].size(), false));
        for (int i = 0; i < grid.size(); i++) {
            for(int j = 0; j < grid[0].size(); j++) {
                ans = max(ans, getArea(grid, i, j, visited));
            }
        }
        return ans;
    }
    
    int getArea(vector<vector<int>>& grid, int row, int col, vector<vector<bool>>& visited) {
        if (row < 0 || row > grid.size() - 1 || col < 0 || col > grid[0].size() - 1 || visited[row][col] || grid[row][col] == 0) {
            return 0;
        }
        visited[row][col] = true;
        return 1 + getArea(grid, row + 1, col, visited) + getArea(grid, row - 1, col, visited) 
                + getArea(grid, row, col + 1, visited) + getArea(grid, row, col - 1, visited);
    }
};

File: golang/lc-695.go
package main

func maxAreaOfIsland(grid [][]int) int {
	res := 0
	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[0]); j++ {
			if grid[i][j] == 1 {
				res = max(res, dfs695(grid, i, j))
			}
		}
	}
	return res
}

func dfs695(grid [][]int, row, col int) int {
	val := 0
	grid[row][col] = 2
	if row < len(grid)-1 && grid[row+1][col] == 1 {
		val += dfs695(grid, row+1, col)
	}

	if col < len(grid[0])-1 && grid[row][col+1] == 1 {
		val += dfs695(grid, row, col+1)
	}
	if row > 0 && grid[row-1][col] == 1 {
		val += dfs695(grid, row-1, col)
	}
	if col > 0 && grid[row][col-1] == 1 {
		val += dfs695(grid, row, col-1)
	}
	return val + 1
}
